pkg/cli/start.go

```diff
@@ -689,7 +689,21 @@ func runStart(cmd *cobra.Command, args []string) error {
 			stopper.Stop(ctx)
 		}()
 
-		// Don't return: we're shutting down gracefully.
+	// Don't return: we're shutting down gracefully.
+
+	case <-log.FatalChan:
+		// A fatal error has occurred. Stop everything (gracelessly) to
+		// avoid serving incorrect data while the final log messages are
+		// being written.
+		// https://github.com/cockroachdb/cockroach/issues/23414
+		// TODO(bdarnell): This could be more graceless, for example by
+		// reaching into the server objects and closing all the
+		// connections while they're in use. That would be more in line
+		// with the expected effect of a log.Fatal.
+		stopper.Stop(ctx)
```

Comment:
This will cause the exit in cobra (which will process the return here) to race with the exit in log.Fatal.

Set `returnErr` here so that both will see the same process exit status (255).